src/ol_infrastructure/components/aws/auto_scale_group.py

```diff
@@ -0,0 +1,345 @@
+        cls: "OLLoadBalancerConfig", load_balancer_type: str
+    ) -> str:
+        if load_balancer_type not in ["application", "gateway", "network"]:
+            raise ValueError(
+                f"load_balancer_type: {load_balancer_type} is not valid. Only 'application', 'gateway', or 'network' are accepted"
+            )
+        return load_balancer_type
+
+
+class OLLaunchTemplateConfig(AWSBase):
+    """Configuration Object for defining configuration needed to create a launch template."""
+
+    block_device_mappings: List[BlockDeviceMapping]
+    image_id: str
+    instance_profile_arn: Union[str, pulumi.Output[str]]
+    instance_type: InstanceTypes = InstanceTypes.burstable_medium
+    key_name: str = "oldevops"
+    security_groups: List[Union[SecurityGroup, pulumi.Output]]
+    tag_specifications: List[TagSpecification]
+    user_data: Optional[Union[str, pulumi.Output[str]]]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+class OLAutoScaleGroupConfig(AWSBase):
+    """Configuration Object for defining configuration needed to create an autoscale group."""
+
+    asg_name: str
+    aws_config: AWSBase
+    desired_size: PositiveInt = PositiveInt(2)
+    health_check_grace_period: NonNegativeInt = NonNegativeInt(0)
+    health_check_type: str = "ELB"
+    max_size: PositiveInt = PositiveInt(2)
+    min_size: PositiveInt = PositiveInt(1)
+    vpc_zone_identifiers: pulumi.Output[str]
+
+    instance_refresh_checkpoint_delay: PositiveInt = PositiveInt(3600)
+    instance_refresh_checkpoint_percentages: List[NonNegativeInt] = []
+    instance_refresh_warmup: PositiveInt = PositiveInt(health_check_grace_period)
+    instance_refresh_min_healthy_percentage: NonNegativeInt = NonNegativeInt(50)
+    instance_refresh_strategy: str = "Rolling"
+    instance_refresh_triggers: List[str] = ["tags"]
+
+    class Config:
+        arbitrary_types_allowed = True
+
+    @validator("instance_refresh_strategy")
+    def is_valid_strategy(cls: "OLAutoScaleGroupConfig", strategy: str) -> str:
+        if strategy != "Rolling":
+            raise ValueError("The only vaild instance refresh strategy is 'Rolling'")
+        return strategy
+
+    @validator("health_check_type")
+    def is_valid_healthcheck(
+        cls: "OLAutoScaleGroupConfig", health_check_type: str
+    ) -> str:
+        if health_check_type not in ["ELB", "EC2"]:
+            raise ValueError(
+                f"health_check_type: {health_check_type} is not valid. Only 'ELB' or 'EC2' are accepted."
+            )
+        return health_check_type
+
+
+class OLAutoScaling(pulumi.ComponentResource):
+    """Component to create an autoscaling group with defaults as well as managed associated resources"""
+
+    lb_dns_name: str
+
+    def __init__(
+        self,
+        asg_config: OLAutoScaleGroupConfig,
+        lt_config: OLLaunchTemplateConfig,
+        tg_config: OLTargetGroupConfig,
+        lb_config: OLLoadBalancerConfig,
```

Comment:
Should we make this `Optional[OLLoadBalancerConfig]` and then key off of that to determine whether to create the load balancer? That way we can easily use the same component for ASGs that do and don't want an LB

Reply:
That makes sense. I didn't know there was a use case that omits the LB.